### PR TITLE
Formatted Time for Display on Edit

### PR DESCRIPTION
**Pull request by Boorclark**
Fixes Issue #355 

Changes Made:
 - Reformatted time picker HTML to the same format as a date picker for clarity.
 - Created a function in utils for formatting 24 hour time (str)  --> 12 hour (str) time (in utils for future use)
 - Added conditional in routes so the time is only formatted if the user does not have the privilege to edit the event.

--------------------------------------------------------------------------------------------------------------------------------------------

image
image

--------------------------------------------------------------------------------------------------------------------------------------------

Files Changed:
 - app/controllers/admin/routes.py
 - app/logic/utils.py
 - app/templates/admin/createSingleEvent.html

### CODE REVIEW

File: app/controllers/admin/routes.py [hunk 1]
```diff
@@ -141,8 +140,12 @@ def editEvent(eventId):
     userHasRSVPed = EventRsvp.get_or_none(EventRsvp.user == g.current_user, EventRsvp.event == event)
     isPastEvent = (datetime.now() >= datetime.combine(event.startDate, event.timeStart))
     program = event.singleProgram
-
     isProgramManager = hasPrivilege(g.current_user,program)
+
+    if (g.current_user.isStudent or g.current_user.isFaculty) and not isProgramManager:  # only formats to 12 hour time if user doesn't have access (for display purposes)
+        eventData['timeStart'] = format24to12HourTime(eventData['timeStart'])
+        eventData['timeEnd'] = format24to12HourTime(eventData['timeEnd'])
```
Comment (BrianRamsay) [inline app/controllers/admin/routes.py]: Doesn't the time picker require 24 hour time? I thought that was the whole issue you dealt with in the other timepicker PR

**Review by BrianRamsay — CHANGES_REQUESTED** [review]

File: app/logic/utils.py [hunk 2]
```diff
@@ -58,7 +58,20 @@ def format24HourTime(unformattedTime):
             #  calling strptime here to explicitly raise an exception if it wasn't properly in 24 hour format
             formattedTime = datetime.strptime(unformattedTime, "%H:%M")
             return unformattedTime
-    elif isinstance(unformattedTime, datetime):
+    else:
```
Comment (BrianRamsay) [inline app/logic/utils.py:61]: good change. leave as is

File: app/logic/utils.py [hunk 3]
```diff
@@ -58,7 +58,20 @@ def format24HourTime(unformattedTime):
...
+    :returns: Formatted time string (12 hour time)
+    """
+    if int(timeStr[:2]) > 12:
+      formattedTime = str(int(timeStr[:2]) - 12) + str(timeStr[2:]) + " PM"
+    elif int(timeStr[:2]) < 12:
+      formattedTime =  timeStr + " AM"
+    else:
+      formattedTime = timeStr + " PM"
+
+    return formattedTime;
```
Comment (BrianRamsay) [inline app/logic/utils.py]: Replace with `datetime.strptime("13:47", "%H:%M").strftime("%I:%M %p")`

File: app/templates/admin/createSingleEvent.html [hunk 4]
```diff
@@ -111,24 +111,22 @@ <h1 id="title">{{page_title}}</h1>
...
               <div class="input-group">
-                {% set startTime = eventData.timeStart.strftime('%H:%M') if eventData.timeStart and eventData.timeStart.strftime else eventData.timeStart%}
-                <input autocomplete="off" type="time" class="form-control timepicker" id='startTime' value="{{startTime}}" name="timeStart" aria-describedby="timeIcon1" placeholder="Pick a start time" required />
+                <input autocomplete="off" type="time" class="form-control timepicker readonly" value="{{ eventData.timeStart.strftime('%H:%M') if eventData.timeStart and eventData.timeStart.strftime else eventData.timeStart }}" name="timeStart" aria-describedby="timeIcon1" placeholder="Pick a start time" onchange="update(this)" id=startTime required />
                 <span class="input-group-text timeIcons" id="timeIcon1" hidden><i class="bi bi-clock"></i></span>
               </div>
               {% else %}
-                <p>{{eventData.timeStart.strftime('%H:%M') if eventData.timeStart }}</p>
+                <p>{{eventData.timeStart.strftime('%H:%M') if eventData.timeStart and eventData.timeStart.strftime else eventData.timeStart}}</p>
```
Comment (BrianRamsay) [inline app/templates/admin/createSingleEvent.html]: We shouldn't need the formatting anymore, just output the text. Double check this assumption

**Review by BrianRamsay — CHANGES_REQUESTED** [review]

File: app/controllers/admin/routes.py (hunk 1 above)
Comment (Boorclark) [reply app/controllers/admin/routes.py]: Yes the time picker does, not the text for people w/ out edit access

File: app/logic/utils.py (hunk 3 above)
Comment (Boorclark) [reply app/logic/utils.py]: resolved

File: app/templates/admin/createSingleEvent.html (hunk 4 above)
Comment (Boorclark) [reply app/templates/admin/createSingleEvent.html]: resolved

Comment (BrianRamsay) [reply app/templates/admin/createSingleEvent.html]: Didn't we say we don't need the conditional anymore?

Comment (BrianRamsay) [reply app/templates/admin/createSingleEvent.html]: eventData.timeStart is always a string

**Review by BrianRamsay — CHANGES_REQUESTED** [review]

**Review by BrianRamsay — APPROVED** [review]

**Pull request merged**